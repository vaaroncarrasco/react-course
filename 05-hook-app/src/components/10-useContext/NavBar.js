import React from 'react'
import { NavLink } from 'react-router-dom';

// Link -> like a
// NavLink -> like a but with exact & activeClassName attribute

export const NavBar = () => {
    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">

            <NavLink to="/" className="navbar-brand">useContext</NavLink>

            <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                <div className="navbar-nav">
                    <NavLink exact activeClassName="active" to="/" className="nav-item nav-link">Home</NavLink>
                    <NavLink exact activeClassName="active" to="/about" className="nav-item nav-link">About</NavLink>
                    <NavLink exact activeClassName="active" to="/login" className="nav-item nav-link">Login</NavLink>
                </div>

            </div>

        </nav>
    )
}
