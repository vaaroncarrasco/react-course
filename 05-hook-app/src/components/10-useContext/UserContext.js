import { createContext } from 'react';

// * Higher Order Component

// 1. createContext()
export const UserContext = createContext(null);