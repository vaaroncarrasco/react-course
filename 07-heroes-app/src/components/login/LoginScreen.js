import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext'
import { types } from '../../types/types';

// ? This component inherits props from the router it is called from: like props{}.history
export const LoginScreen = ({ history }) => {

    const { dispatch } = useContext(AuthContext);

    // import context, in handleLogin dispatch action { name: 'fernando' } with type login

    const handleLogin = () => {
        // history.push('/'); // * sends to other direction and user is able to go <- one page back

        const lastPath = localStorage.getItem('lastPath') || '/';

        dispatch({
            type: types.login,
            payload: {
                name: 'Fernando'
            }
        });

        history.replace( lastPath ); // * sends to other direction replacing current page unabling going <- back // good for security
    }

    return (
        <div className="container mt-5">
            <h1>LoginScreen</h1>
            <hr />

            <button
                className="btn btn-primary"
                onClick={ handleLogin }
            >
                Login
            </button>
        </div>
    )

}