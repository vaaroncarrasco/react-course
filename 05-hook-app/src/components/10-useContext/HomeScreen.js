import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomeScreen = () => {

    // 3. useContext -> use createdContext to get data -> import context component and useContext()
    const { user } = useContext(UserContext);

    return (
        <div>
            <h1>HomeScreen</h1>
            <hr />

            <pre>
                {
                    JSON.stringify( user, null, 3)
                }
            </pre>
        </div>
    )
}
