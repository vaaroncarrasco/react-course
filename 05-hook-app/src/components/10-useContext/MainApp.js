import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'

export const MainApp = () => {

    const [user, setUser] = useState({});

    return (
        // * Higher Order Component creates -> Context.Provider -> allowing to pass data thru child components
        // 2. Send data thru context component.Provider value={ value }
        <UserContext.Provider value={{
            user,
            setUser
        }}>

            <AppRouter />

        </UserContext.Provider>
    )
}
