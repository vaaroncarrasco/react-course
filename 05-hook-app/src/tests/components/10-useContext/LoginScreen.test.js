import React from "react";
import { mount } from "enzyme";
import { LoginScreen } from "../../../components/10-useContext/LoginScreen";
import { UserContext } from "../../../components/10-useContext/UserContext";

describe('testing <LoginScreen />', () => {

    const setUser = jest.fn();

    // * Testing useContext -> render w/ mount() wrapping child component with UserContext.Provider/HigherOrderComponent
    const wrapper = mount(
        <UserContext.Provider value={ {setUser} }>
            <LoginScreen />
        </UserContext.Provider>
    );

    test('should match snapshot', () => { expect( wrapper ).toMatchSnapshot(); });

    test('should run setUser w/ expected argument', () => {

        wrapper.find('button').prop('onClick')();

        expect( setUser ).toHaveBeenCalledWith({
            id: 123,
            name: 'Aaron'
        });

    });

});
