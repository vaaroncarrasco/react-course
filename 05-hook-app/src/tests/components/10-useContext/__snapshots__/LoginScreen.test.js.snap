// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <LoginScreen /> should match snapshot 1`] = `
<div>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Log in
  </button>
</div>
`;
